New packager: zypper

jaredd opened the pull request:
The openSUSE distribution uses zypper as its packager.  Zypper is not currently supported by pybombs.  This pull request addresses that deficiency.

Zypper operates on rpms and thus the code was based on that from Redhat's yum packager.  One wrinkle is that zypper is very slow when searching for packages.  The rpm utility that is also included by default on openSUSE can find installed packages almost instantly,  so it is used when available.  Zypper is still used to identify packages and versions for those not installed.  I was able to use this modified pybombs on my own machine to successfully install some dependencies for openSUSE.  

Since openSUSE has a different naming convention than Redhat, I will also issue a pull-request for gr-recipes with openSUSE conformant packages or'd in the rpm satisfy sub-field.

mbr0wn commented:
Whoooooooooooooo! Let me review.

mbr0wn commented:
@jaredd https://travis-ci.org/gnuradio/pybombs/jobs/215785670

Can you please make sure the tests don't fail? You can run them locally using ./run-tests.sh. Also, if you're feeling extra generous, you could add a Docker test for OpenSUSE, but I won't wait with merging because of this.

jaredd commented:
Martin,

Ugh, I tried a different editor while developing and I guess I didn't check
this.  Reran run_tests locally:

$ ./run-tests.sh
Running static code analysis (PyLint)...
Problem importing module imports.py: No module named isort
Problem importing module variables.py: No module named
backports.functools_lru_cache
No config file found, using default configuration
Building source distribution package (sdist)...
warning: sdist: standard file not found: should have one of README,
README.rst, README.txt

So I think that fixed the issue.  Do I need to create a new pull request or
can you just pull from the latest off my repo branch?
Jared.

On Tue, Mar 28, 2017 at 7:42 AM, Martin Braun <notifications@github.com>
wrote:

> @jaredd <https://github.com/jaredd> https://travis-ci.org/
> gnuradio/pybombs/jobs/215785670
>
> Can you please make sure the tests don't fail? You can run them locally
> using ./run-tests.sh. Also, if you're feeling extra generous, you could add
> a Docker test for OpenSUSE, but I won't wait with merging because of this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gnuradio/pybombs/pull/429#issuecomment-289792418>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADFk2zTq6lvjblAlBFzmzjt-ezwtICI_ks5rqRxNgaJpZM4MrGUQ>
> .
>

jaredd commented:
Do you have an example of a docker test for another packager?  Maybe I
could figure it out.

On Tue, Mar 28, 2017 at 8:39 PM, U L <jareddpub@gmail.com> wrote:

> Martin,
>
> Ugh, I tried a different editor while developing and I guess I didn't
> check this.  Reran run_tests locally:
>
> $ ./run-tests.sh
> Running static code analysis (PyLint)...
> Problem importing module imports.py: No module named isort
> Problem importing module variables.py: No module named
> backports.functools_lru_cache
> No config file found, using default configuration
> Building source distribution package (sdist)...
> warning: sdist: standard file not found: should have one of README,
> README.rst, README.txt
>
> So I think that fixed the issue.  Do I need to create a new pull request
> or can you just pull from the latest off my repo branch?
> Jared.
>
> On Tue, Mar 28, 2017 at 7:42 AM, Martin Braun <notifications@github.com>
> wrote:
>
>> @jaredd <https://github.com/jaredd> https://travis-ci.org/gnuradio
>> /pybombs/jobs/215785670
>>
>> Can you please make sure the tests don't fail? You can run them locally
>> using ./run-tests.sh. Also, if you're feeling extra generous, you could add
>> a Docker test for OpenSUSE, but I won't wait with merging because of this.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/gnuradio/pybombs/pull/429#issuecomment-289792418>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADFk2zTq6lvjblAlBFzmzjt-ezwtICI_ks5rqRxNgaJpZM4MrGUQ>
>> .
>>
>
>

mbr0wn commented:
Checks are auto-run, and they passed :)

Look at `tests/docker` for other examples.

mbr0wn commented:
Sorry about those `{}` comments, I meant to post all of the comments at once, but I hadn't submitted the code review.

mbr0wn commented on pybombs/packagers/zypper.py:
```diff
@@ -0,0 +1,170 @@
+            # So, two steps:
+            # 1) Check that pkgname is correct (and, if necessary, the package arch)
+            # 2) return version
+            match_pat = r"^(?P<status>)\s+\|\s+(?P<pkg>[^\.]+)\s+\| \
```
Will this regex work if the user is running OpenSUSE with a different locale? We had some trouble with the Chinese version of CentOS. See the rpm packager on how to fix that.

jaredd commented:
Martin,

Thanks for the review.  The rpm tool can only find installed packages,
zypper can find those available in a remote repository.  But since the
function itself is only interested in what version is installed, then seems
the zypper fallback wouldn't be necessary.  So I have removed the comment.

I have no experience with internationalization.  I assume when you say
"rpm" packager you mean yum.  There's a line there
<https://github.com/gnuradio/pybombs/blob/master/pybombs/packagers/yum.py#L54>that
seems to refer to some locale related environment variable.  I've added
that to my zypper packager aping the match_output call in yum.

I've also changed the format string replacement fields per your suggestion.

The docker stuff I'm looking at.  I don't understand the syntax yet.  One
issue I have is that I can't test with the master branch of the
gnuradio/gr-recipes repo.  I need to use one of my (non-master) branches
that has the opensuse package names.  Can you recommend a line I can
include in my dockerfile that sets the gr-recipes branch?  I've modeled it
on the fedora25 dockerfile, replacing dnf with zypper, changing the package
name, and adjusting the FROM line to reflect opensuse (tumbleweed, which is
what I'm running).

Thanks,
Jared.

On Wed, Mar 29, 2017 at 3:59 PM, Martin Braun <notifications@github.com>
wrote:

> *@mbr0wn* commented on this pull request.
> ------------------------------
>
> In pybombs/packagers/zypper.py
> <https://github.com/gnuradio/pybombs/pull/429#discussion_r108807632>:
>
> > +            except Exception as ex:
> +                self.log.error("Parsing `{0} list installed` failed.".format(self.fastcommand))
> +                self.log.obnoxious(str(ex))
> +            return False    # should we fall back to zypper instead?
> +        try:
> +            # 'list installed' will return non-zero if package does not exist, thus will throw
> +            out = subproc.check_output(
> +                    [self.command, "search", "-is", pkgname],
> +                    stderr=subprocess.STDOUT
> +            ).strip().split("\n")
> +            # Output looks like this:
> +            # <status>|<pkgname>|<type>|<version>|<arch>|<repo>
> +            # So, two steps:
> +            # 1) Check that pkgname is correct (and, if necessary, the package arch)
> +            # 2) return version
> +            match_pat = r"^(?P<status>)\s+\|\s+(?P<pkg>[^\.]+)\s+\| \
>
> Will this regex work if the user is running OpenSUSE with a different
> locale? We had some trouble with the Chinese version of CentOS. See the rpm
> packager on how to fix that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gnuradio/pybombs/pull/429#pullrequestreview-29865227>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADFk29Wb39zidRTxW_GEUj3luoC8HcgQks5rquJtgaJpZM4MrGUQ>
> .
>

mbr0wn commented:
You can use
    $ pybombs config --package gr-foo source http://git.foo.bar/...

mbr0wn commented:
Awesome, thanks!

jaredd commented:
Is this going to work for recipes also?  Would I do something like

$ pybombs config --package gr-recipes gitbranch my-custom-branch

?

On Thu, Mar 30, 2017 at 10:17 PM, Martin Braun <notifications@github.com>
wrote:

> You can use
> $ pybombs config --package gr-foo source http://git.foo.bar/...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gnuradio/pybombs/pull/429#issuecomment-290618315>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADFk26F4rAEfKc62PCjPRIuKybraZ34Jks5rrIyAgaJpZM4MrGUQ>
> .
>
